From 654f5f786c1f292d3777cf99db81f415525e61a8 Mon Sep 17 00:00:00 2001
From: wangyunhua <wangyunhua3@contributor>


Date: Tue, 10 Feb 2026 02:45:45 +0000
Subject: [PATCH] BUGFIX  Fix crash in foreign key cases with table
 recycle bin enabled.

Issue:
------
When the table recycle bin feature is disabled, dropping a child table
only removes it from the `cde_dict_t` map. When the feature is enabled
and foreign keys are involved, both the child and parent tables are
removed from `cde_dict_t`. Later access to these tables requires reopening
them. During the reopening of the parent table, the `mdl` pointer is not
cleared after the first iteration, leading to a crash in the second iteration.

Solutions:
---------
Clear the `mdl` pointer after each iteration.
---
 .../r/dstore-index-online-fk.result           | 26 ++++++++++++++
 .../dstore_main/t/dstore-index-online-fk.test | 34 +++++++++++++++++++
 storage/dstore/ddl/ha_cde_alter.cc            |  2 +-
 3 files changed, 61 insertions(+), 1 deletion(-)

diff --git a/mysql-test/suite/dstore_main/r/dstore-index-online-fk.result b/mysql-test/suite/dstore_main/r/dstore-index-online-fk.result
index c39ca6b56d5..02a5e2f7160 100644
--- a/mysql-test/suite/dstore_main/r/dstore-index-online-fk.result
+++ b/mysql-test/suite/dstore_main/r/dstore-index-online-fk.result
@@ -422,3 +422,29 @@ SET DEBUG = '+d,dstore_inplace_alter_database_is_nullptr';
 ALTER TABLE t1 ADD FOREIGN KEY (b) REFERENCES non(a);
 SET DEBUG = '-d,dstore_inplace_alter_database_is_nullptr';
 DROP TABLE t1;
+# Bug #BUGFIX	 crashes after enabling the table recycle bin.
+# WITH rds_recycle_bin_mode ON
+SET rds_recycle_bin_mode = ON;
+CREATE TABLE parent (a INT PRIMARY KEY, b INT NOT NULL, c INT) ENGINE = dstore;
+INSERT INTO parent VALUES(1,2,3),(2,3,4);
+CREATE INDEX tb ON parent(b);
+CREATE TABLE child (a1 INT NOT NULL, a2 INT, a3 INT) ENGINE = dstore;
+CREATE INDEX tb ON child(a2);
+ALTER TABLE child
+ADD CONSTRAINT fk_a FOREIGN KEY (a2) REFERENCES parent(b)
+ON DELETE SET NULL ON UPDATE CASCADE,
+ALGORITHM = INPLACE;
+ALTER TABLE child
+ADD CONSTRAINT fk_b FOREIGN KEY (a1) REFERENCES parent(a),
+ALGORITHM = INPLACE;
+DROP TABLE child;
+CREATE TABLE child (a1 INT NOT NULL, a2 INT, a3 INT) ENGINE = dstore;
+CREATE INDEX tb ON child(a2);
+ALTER TABLE child ADD PRIMARY KEY idx (a1),
+ADD CONSTRAINT fk_new_1 FOREIGN KEY (a1) REFERENCES parent(b),
+ADD CONSTRAINT fk_new_2 FOREIGN KEY (a2) REFERENCES parent(a),
+ADD CONSTRAINT fk_new_3 FOREIGN KEY (a3) REFERENCES parent(a),
+ALGORITHM = INPLACE;
+DROP TABLE child;
+DROP TABLE parent;
+SET rds_recycle_bin_mode = OFF;
diff --git a/mysql-test/suite/dstore_main/t/dstore-index-online-fk.test b/mysql-test/suite/dstore_main/t/dstore-index-online-fk.test
index a77ce8bcfc0..be6e3df752d 100644
--- a/mysql-test/suite/dstore_main/t/dstore-index-online-fk.test
+++ b/mysql-test/suite/dstore_main/t/dstore-index-online-fk.test
@@ -474,3 +474,37 @@ SET DEBUG = '+d,dstore_inplace_alter_database_is_nullptr';
 ALTER TABLE t1 ADD FOREIGN KEY (b) REFERENCES non(a);
 SET DEBUG = '-d,dstore_inplace_alter_database_is_nullptr';
 DROP TABLE t1;
+
+--echo # Bug #BUGFIX	 crashes after enabling the table recycle bin.
+--echo # WITH rds_recycle_bin_mode ON
+SET rds_recycle_bin_mode = ON;
+
+CREATE TABLE parent (a INT PRIMARY KEY, b INT NOT NULL, c INT) ENGINE = dstore;
+INSERT INTO parent VALUES(1,2,3),(2,3,4);
+CREATE INDEX tb ON parent(b);
+
+CREATE TABLE child (a1 INT NOT NULL, a2 INT, a3 INT) ENGINE = dstore;
+CREATE INDEX tb ON child(a2);
+
+ALTER TABLE child
+ADD CONSTRAINT fk_a FOREIGN KEY (a2) REFERENCES parent(b)
+ON DELETE SET NULL ON UPDATE CASCADE,
+ALGORITHM = INPLACE;
+
+ALTER TABLE child
+ADD CONSTRAINT fk_b FOREIGN KEY (a1) REFERENCES parent(a),
+ALGORITHM = INPLACE;
+
+DROP TABLE child;
+CREATE TABLE child (a1 INT NOT NULL, a2 INT, a3 INT) ENGINE = dstore;
+CREATE INDEX tb ON child(a2);
+
+ALTER TABLE child ADD PRIMARY KEY idx (a1),
+ADD CONSTRAINT fk_new_1 FOREIGN KEY (a1) REFERENCES parent(b),
+ADD CONSTRAINT fk_new_2 FOREIGN KEY (a2) REFERENCES parent(a),
+ADD CONSTRAINT fk_new_3 FOREIGN KEY (a3) REFERENCES parent(a),
+ALGORITHM = INPLACE;
+
+DROP TABLE child;
+DROP TABLE parent;
+SET rds_recycle_bin_mode = OFF;
diff --git a/storage/dstore/ddl/ha_cde_alter.cc b/storage/dstore/ddl/ha_cde_alter.cc
index c3df94dbcec..3ef636e6cf7 100644
--- a/storage/dstore/ddl/ha_cde_alter.cc
+++ b/storage/dstore/ddl/ha_cde_alter.cc
@@ -2017,7 +2017,6 @@ added
   char *referenced_table_name = nullptr;
   uint32_t num_fk = 0;
   Alter_info *alter_info = ha_alter_info->alter_info;
-  MDL_ticket *mdl{nullptr};
 
   DBUG_TRACE;
 
@@ -2025,6 +2024,7 @@ added
 
   constexpr uint32_t MAX_NUM_FK_COLUMNS = 500;
   for (const Key_spec *key : alter_info->key_list) {
+    MDL_ticket *mdl{nullptr};
     if (key->type != KEYTYPE_FOREIGN) {
       continue;
     }
-- 
2.23.0

